rbd: VolumeGroupReplicationContent controller to regenerate the OMAP data

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This commit adds new controller that watches for the
VolumeGroupReplicationContent and regenerates the OMAP data if
it doesn't exists.


**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [x] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
rbd: controller to regenerate volume group omap data

This commit adds new controller that watches for the
VolumeGroupReplicationContent and regenerates the OMAP data if
it doesn't exists.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar committed:
journal: pass groupUUID to be used for omap name reserve

This commit adds groupUUID param for `ReserveName` to be used for
OMAP name reserve instead of auto-generating.
This is useful for mirroring and metro-DR ensuring that mirrored
resources have consistent OMAP names across mirrored clusters.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/rbd/group/volume_group.go:
```diff
@@ -35,16 +35,16 @@ import (
-type volumeGroup struct {
-	// id is a unique value for this volume group in the Ceph cluster, it
+// VolumeGroup handles all requests for 'rbd group' operations.
+type VolumeGroup struct {
```
Don't export these, they are private intentional. They implement the interface in the `internal/rbd/types/*.go` files. You can use `rbd.NewManager()` to get a manager struct and use the Manager API to get the VolumeGroup.

iPraveenParihar replied:
If you are referring to `GetVolumeGroupByID`, then It doesn't suite well to use it here. Because in `GetVolumeGroup` we have GetPoolName by `LocationID` (which is PoolID) and PoolID may be different in secondary cluster. 

https://github.com/ceph/ceph-csi/blob/130e8b42f2ef72bd54e8dc0f96823af3c3890ee8/internal/rbd/group/volume_group.go#L106-L109

Else we can do it by passing an optional parameter `PoolName` for `GetVolumeGroup` ?
If it is passed as empty, then call `GetPoolName` else use the `PoolName` passed?

nixpanic replied:
You may want to add more functions to the manager and other types. Ideally the API that the manager provides stays simple to use.

iPraveenParihar commented:
#### Initial Test results - 
    
```
E0806 10:45:48.804306       1 omap.go:80] omap not found (pool="replicapool", namespace="", name="csi.groups.default"): rados: ret=-2, No such file or directory
I0806 10:45:48.804338       1 rbd_journal.go:764] the journal does not contain a reservation for a volume group with name "vgrcontent-88b1d055-4c66-4218-a81c-487ba0b58ccf" yet
I0806 10:45:48.811316       1 omap.go:159] set omap keys (pool="replicapool", namespace="", name="csi.groups.default"): map[csi.volume.group.vgrcontent-88b1d055-4c66-4218-a81c-487ba0b58ccf:88b1d055-4c66-4218-a81c-487ba0b58ccf])
I0806 10:45:48.814878       1 omap.go:159] set omap keys (pool="replicapool", namespace="", name="csi.volume.group.88b1d055-4c66-4218-a81c-487ba0b58ccf"): map[csi.groupname:csi-vol-group-88b1d055-4c66-4218-a81c-487ba0b58ccf csi.volname:vgrcontent-88b1d055-4c66-4218-a81c-487ba0b58ccf])
I0806 10:45:48.816837       1 omap.go:89] got omap values: (pool="replicapool", namespace="", name="csi.volume.5bfefce1-6427-45a3-b088-6d43d19b1e9c"): map[csi.imageid:288758945f2b8 csi.imagename:csi-vol-5bfefce1-6427-45a3-b088-6d43d19b1e9c csi.volname:pvc-34dc4369-2eb3-492a-88aa-5cd37b557154 csi.volume.owner:rook-ceph]
I0806 10:45:48.838187       1 omap.go:89] got omap values: (pool="replicapool", namespace="", name="csi.volume.896c6912-604b-43db-916c-1e386a71973f"): map[csi.imageid:216b4d44baf60 csi.imagename:csi-vol-896c6912-604b-43db-916c-1e386a71973f csi.volname:pvc-2506eb3c-69e9-4054-8b8d-405de060c0f4 csi.volume.owner:rook-ceph]
I0806 10:45:48.856998       1 rbd_journal.go:799] adding volume mapping for volume "0001-0009-rook-ceph-0000000000000001-5bfefce1-6427-45a3-b088-6d43d19b1e9c" to volume group "csi-vol-group-88b1d055-4c66-4218-a81c-487ba0b58ccf"
I0806 10:45:48.860180       1 omap.go:159] set omap keys (pool="replicapool", namespace="", name="csi.volume.group.88b1d055-4c66-4218-a81c-487ba0b58ccf"): map[0001-0009-rook-ceph-0000000000000001-5bfefce1-6427-45a3-b088-6d43d19b1e9c:])
I0806 10:45:48.860287       1 rbd_journal.go:799] adding volume mapping for volume "0001-0009-rook-ceph-0000000000000001-896c6912-604b-43db-916c-1e386a71973f" to volume group "csi-vol-group-88b1d055-4c66-4218-a81c-487ba0b58ccf"
I0806 10:45:48.863734       1 omap.go:159] set omap keys (pool="replicapool", namespace="", name="csi.volume.group.88b1d055-4c66-4218-a81c-487ba0b58ccf"): map[0001-0009-rook-ceph-0000000000000001-896c6912-604b-43db-916c-1e386a71973f:])
I0806 10:45:48.863972       1 rbd_journal.go:816] re-generated Group ID (0001-0009-rook-ceph-0000000000000001-88b1d055-4c66-4218-a81c-487ba0b58ccf) and Group Name (csi-vol-group-88b1d055-4c66-4218-a81c-487ba0b58ccf) for request name (vgrcontent-88b1d055-4c66-4218-a81c-487ba0b58ccf)
```
    
```
bash-5.1$ rados listomapvals csi.groups.default -p replicapool
csi.volume.group.vgrcontent-88b1d055-4c66-4218-a81c-487ba0b58ccf
value (36 bytes) :
00000000  38 38 62 31 64 30 35 35  2d 34 63 36 36 2d 34 32  |88b1d055-4c66-42|
00000010  31 38 2d 61 38 31 63 2d  34 38 37 62 61 30 62 35  |18-a81c-487ba0b5|
00000020  38 63 63 66                                       |8ccf|
00000024

bash-5.1$ rados listomapvals csi.volume.group.88b1d055-4c66-4218-a81c-487ba0b58ccf -p replicapool
0001-0009-rook-ceph-0000000000000001-5bfefce1-6427-45a3-b088-6d43d19b1e9c
value (0 bytes) :

0001-0009-rook-ceph-0000000000000001-896c6912-604b-43db-916c-1e386a71973f
value (0 bytes) :

csi.groupname
value (50 bytes) :
00000000  63 73 69 2d 76 6f 6c 2d  67 72 6f 75 70 2d 38 38  |csi-vol-group-88|
00000010  62 31 64 30 35 35 2d 34  63 36 36 2d 34 32 31 38  |b1d055-4c66-4218|
00000020  2d 61 38 31 63 2d 34 38  37 62 61 30 62 35 38 63  |-a81c-487ba0b58c|
00000030  63 66                                             |cf|
00000032

csi.volname
value (47 bytes) :
00000000  76 67 72 63 6f 6e 74 65  6e 74 2d 38 38 62 31 64  |vgrcontent-88b1d|
00000010  30 35 35 2d 34 63 36 36  2d 34 32 31 38 2d 61 38  |055-4c66-4218-a8|
00000020  31 63 2d 34 38 37 62 61  30 62 35 38 63 63 66     |1c-487ba0b58ccf|
```

nixpanic commented on internal/rbd/manager.go:
```diff
@@ -48,7 +48,7 @@ type rbdManager struct {
 // NewManager returns a new manager for handling Volume and Volume Group
 // operations, combining the requests for RBD and the journalling in RADOS.
-func NewManager(csiID string, parameters, secrets map[string]string) types.Manager {
+func NewManager(csiID string, parameters, secrets map[string]string) *rbdManager {
```
no, do not do this. You can extend the `types.Manager` interface if you need more functions.

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -701,3 +702,145 @@ func (rv *rbdVolume) storeImageID(ctx context.Context, j *journal.Connection) er
+	volumeIds []string,
+	parameters,
+	secrets map[string]string,
+) (string, error) {
```
what is this `string` that gets returned? Please mention it in the comment.

nixpanic commented on internal/rbd/rbd_journal.go:
```diff
@@ -701,3 +702,145 @@ func (rv *rbdVolume) storeImageID(ctx context.Context, j *journal.Connection) er
+	)
+	ctx := context.Background()
+
+	mgr, ok := (NewManager(instanceID, parameters, secrets)).(*rbdManager)
```
Why not add `RegenerateVolumeGroupJournal` as a new function to the `types.Manager` interface?

You would then use it like
```go
mgr := rbd.NewManager(instanceID, parameters, secrets)
str, err := mgr.RegenerateVolumeGroupJournal(ctx, groupID, requestName, clusterName, volumeIds)
```

Inside this function you can directly use all the private attributes and functions of rbdManager.

iPraveenParihar committed:
rbd: add RegenerateVolumeGroupJournal method for Manager interface

This commit adds `RegenerateVolumeGroupJournal` to Manager
interface. RegenerateVolumeGroupJournal regenerate the omap
data for the volume group.

This performs the following operations:
  - extracts clusterID and Mons from the cluster mapping
  - Retrieves pool and journalPool parameters from the VolumeGroupReplicationClass
  - Reserves omap data
  - Add volumeIDs mapping to the reserved volume group omap object
  - Generate new volume group handle

Returns the generated volume group handler.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/rbd/manager.go:
```diff
@@ -294,3 +294,134 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+//   - Retrieves pool and journalPool parameters from the VolumeGroupReplicationClass
+//   - Reserves omap data
+//   - Add volumeIDs mapping to the reserved volume group omap object
+//   - Generate new volume group handler
```
it is called "volume group handle" :smiley:

nixpanic replied:
there are still a few references to "volume group handler" instead of "volume group handle"

nixpanic commented on internal/rbd/manager.go:
```diff
@@ -294,3 +294,134 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+		if err != nil {
+			// free up allocated volumes
+			for _, v := range volumes {
+				v.Destroy(ctx)
```
You always need to destroy all volumes that you  retrieved through `GetVolumeByID()`. The most practical way is to do it like:
```go
volumes := make([]types.Volume, len(volumeIds))
defer func() {
    for _, v := range volumes {
        v.Destroy(ctx)
    }
}()
```

nixpanic commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+)
+
+const (
+	secretNameParameterName      string = "replication.storage.openshift.io/group-replication-secret-name"
```
`string` is not needed here

nixpanic reviewed (commented):
thanks for the change, looks much better to me.

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+	// Watch for changes to VolumeGroupReplicationContent
+	err = c.Watch(source.Kind(
+		mgr.GetCache(),
+		&replicationv1alpha1.VolumeGroupReplicationContent{},
```
what will happen if the CRD is not installed on the cluster. The existing functionality should not break

iPraveenParihar replied:
No, this will not cause existing functionality to break.  it will just give warnings.

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
```
no new line is required.

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+	"errors"
+	"fmt"
+
+	ctrl "github.com/ceph/ceph-csi/internal/controller"
```
As this is a new file change the import order as per https://github.com/ceph/ceph-csi/blob/devel/docs/coding.md#imports

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+) (map[string]string, error) {
+	if name == "" || namespace == "" {
+		errStr := "secret name or secret namespace is empty"
+		log.ErrorLogMsg(errStr)
```
no need to log it here. just return and let the caller log it?

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+	vgrClass := &replicationv1alpha1.VolumeGroupReplicationClass{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: vgrc.Spec.VolumeGroupReplicationClassName}, vgrClass)
+	if err != nil {
+		log.ErrorLogMsg("failed to get volumeGroupReplicationClass %s", err)
```
lets not log it here, return it from here and let the caller log it.

Madhu-1 replied:
same is applicable in other places as well.

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	reqName := vgrc.Name
+	groupHandle := vgrc.Spec.VolumeGroupReplicationHandle
```
add a check for empty `VolumeGroupReplicationHandle`

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// Reconcile reconciles the VolumeGroupReplicationContent object and creates a new omap entries
+// for the volume.
```
for the volume group

Madhu-1 commented on internal/journal/volumegroupjournal.go:
```diff
@@ -308,6 +309,7 @@ held, to prevent parallel operations from modifying the state of the omaps for t
 Input arguments:
   - journalPool: Pool where the CSI journal is stored
   - reqName: Name of the volumeGroupSnapshot request received
+  - groupUUID: UUID need to be reserved instead of auto-generating one (this is useful for mirroring and metro-DR)
```
we dont generate anything in metro-DR it only for rbd mirroring

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -294,3 +294,134 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+
+	pool, ok = mgr.parameters["pool"]
+	if !ok {
+		return "", errors.New("required 'pool' parameter missing in volume attributes")
```
missing in parameters

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -294,3 +294,134 @@ func (mgr *rbdManager) DeleteVolumeGroup(ctx context.Context, vg types.VolumeGro
+		return "", err
+	}
+
+	if vgData != nil && vgData.GroupUUID != "" {
```
do we need a check for vgData.GroupUUID here?

iPraveenParihar replied:
Not needed actually, will remove it.

nixpanic commented on internal/rbd/types/manager.go:
```diff
@@ -47,4 +47,8 @@ type Manager interface {
 	DeleteVolumeGroup(ctx context.Context, vg VolumeGroup) error
+
+	// RegenerateVolumeGroupJournal regenerate the omap data for the volume group.
+	// returns the volume group handler
```
Here too, "volume group handler"

iPraveenParihar replied:
done

nixpanic reviewed (changes requested):
Just a nit, but should get corrected never the less.

iPraveenParihar commented:
@nixpanic, I have addressed your requested changes and rebased.

nixpanic reviewed (dismissed):
Thanks!

Looks like there is a conflict in `go.mod`, that needs addressing before this can be merged.

iPraveenParihar commented:
> Thanks!
> 
> Looks like there is a conflict in `go.mod`, that needs addressing before this can be merged.

Thanks, rebased now

Madhu-1 commented on go.mod:
```diff
@@ -48,8 +48,11 @@ require (
+	github.com/csi-addons/kubernetes-csi-addons v0.9.0
 )
 
+require github.com/go-task/slim-sprig/v3 v3.0.0 // indirect
```
we already have require section for indirect  dep, this should be added there, please check

Rakshith-R replied:
This is still to be done

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -503,3 +503,134 @@ func (mgr *rbdManager) CreateVolumeGroupSnapshot(
+		return "", errors.New("required 'pool' parameter missing in parameters")
+	}
+
+	journalPool = mgr.parameters["journalPool"]
+	if journalPool == "" {
+		journalPool = pool
+	}
```
Do we have journalPool in volumegroup?

iPraveenParihar replied:
CreateVolumeGroup has it  - https://github.com/ceph/ceph-csi/blob/a32ba13045f4f433d4c00f568c67209556428a76/internal/rbd/manager.go#L254-L258

Madhu-1 replied:
okay, i dont know in which case we will have journalPool, i think we removed it as its was not required. cc @nixpanic

iPraveenParihar replied:
@nixpanic thoughts? should we remove `journalPool` from everywhere?

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -503,3 +503,134 @@ func (mgr *rbdManager) CreateVolumeGroupSnapshot(
+	}
+	defer vgJournal.Destroy()
+
+	namePrefix = mgr.parameters["volumeNamePrefix"]
```
do we have `volumeNamePrefix` in group?

iPraveenParihar replied:
actually, it should be `groupNamePrefix`.  CreateVolumeGroupSnapshot uses `groupNamePrefix`. 
But CreateVolumeGroup uses `volumeNamePrefix` 
https://github.com/ceph/ceph-csi/blob/a32ba13045f4f433d4c00f568c67209556428a76/internal/rbd/manager.go#L260-L262

Will change it to `groupNamePrefix` ?

Madhu-1 replied:
yes lets stick to one `groupNamePrefix` for both as they need to work together

iPraveenParihar replied:
I noticed CephFS uses `volumeGroupNamePrefix` naming. Let's keep it consistent across component.
Raised a separate PR to address this - https://github.com/ceph/ceph-csi/pull/5063

Madhu-1 reviewed (commented):
@iPraveenParihar i hope you already tested this change and we are able to regenerate and able to find the group with newly generated omap.

Nikhil-Ladha commented on internal/rbd/manager.go:
```diff
@@ -503,3 +503,134 @@ func (mgr *rbdManager) CreateVolumeGroupSnapshot(
+		return "", fmt.Errorf("failed to get poolID for %q: %w", groupUUID, err)
+	}
+
+	groupHandle, err := util.GenerateVolID(ctx, monitors, mgr.creds, poolID, pool, clusterID, groupUUID)
```
Will it return an updated handle when we modify the VolumeGroup? If yes, then we need to check for the VG response always while calling the ModifyVolumeGroup RPC and update the VGRContent CR as well, if the handle has changed.

Nikhil-Ladha replied:
Based on the params passed to the function, I am assuming if the handle is already created it should not get updated as the params would remain the same 🤔

iPraveenParihar replied:
> Based on the params passed to the function, I am assuming if the handle is already created it should not get updated as the params would remain the same

It returns an updated volumeGroupHandle if the `clusterID` / `poolID` or also if the object `UUID` is different in mirrored cluster. Here, we just log it but we don't update the volumeGroupHandle in the `VolumeGroupReplicationContent` CR.

iPraveenParihar committed:
helm: add RBACS for replication.storage.openshift.io

VolumeGroupReplicationContent controller needs `get`, `list` and `watch`
access control for resource `VolumeGroupReplicationContents`. And `get`
access control for resource `VolumeGroupReplicationClasses`.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
@Madhu-1, tested with RBD namespace as well - mirrored namespace `namespace-a`

- primary cluster
```
bash-5.1$ rados listomapvals csi.groups.default -p replicapool -N namespace-a
csi.volume.group.vgrcontent-49fd4502-c600-4922-9317-82906f511325
value (36 bytes) :
00000000  30 39 39 62 61 64 31 38  2d 31 33 32 33 2d 34 64  |099bad18-1323-4d|
00000010  63 65 2d 62 36 61 30 2d  37 39 30 32 65 31 61 31  |ce-b6a0-7902e1a1|
00000020  63 36 66 30                                       |c6f0|
00000024

bash-5.1$ rados listomapvals csi.volume.group.099bad18-1323-4dce-b6a0-7902e1a1c6f0 -p replicapool -N namespace-a
0001-0009-rook-ceph-0000000000000002-2c29fc2c-2888-4aea-8818-b88df1a439e9
value (0 bytes) :

csi.creationtime
value (30 bytes) :
00000000  32 30 32 35 2d 30 31 2d  32 30 54 31 35 3a 35 37  |2025-01-20T15:57|
00000010  3a 30 31 2e 37 35 39 31  38 30 36 30 38 5a        |:01.759180608Z|
0000001e

csi.groupname
value (50 bytes) :
00000000  63 73 69 2d 76 6f 6c 2d  67 72 6f 75 70 2d 30 39  |csi-vol-group-09|
00000010  39 62 61 64 31 38 2d 31  33 32 33 2d 34 64 63 65  |9bad18-1323-4dce|
00000020  2d 62 36 61 30 2d 37 39  30 32 65 31 61 31 63 36  |-b6a0-7902e1a1c6|
00000030  66 30                                             |f0|
00000032

csi.volname
value (47 bytes) :
00000000  76 67 72 63 6f 6e 74 65  6e 74 2d 34 39 66 64 34  |vgrcontent-49fd4|
00000010  35 30 32 2d 63 36 30 30  2d 34 39 32 32 2d 39 33  |502-c600-4922-93|
00000020  31 37 2d 38 32 39 30 36  66 35 31 31 33 32 35     |17-82906f511325|
0000002f
```

- secondary cluster 
```
bash-5.1$ rados listomapvals csi.groups.default -p replicapool -N namespace-a
csi.volume.group.vgrcontent-49fd4502-c600-4922-9317-82906f511325
value (36 bytes) :
00000000  30 39 39 62 61 64 31 38  2d 31 33 32 33 2d 34 64  |099bad18-1323-4d|
00000010  63 65 2d 62 36 61 30 2d  37 39 30 32 65 31 61 31  |ce-b6a0-7902e1a1|
00000020  63 36 66 30                                       |c6f0|
00000024

bash-5.1$ rados listomapvals csi.volume.group.099bad18-1323-4dce-b6a0-7902e1a1c6f0 -p replicapool -N namespace-a
0001-0009-rook-ceph-0000000000000002-2c29fc2c-2888-4aea-8818-b88df1a439e9
value (0 bytes) :

csi.creationtime
value (30 bytes) :
00000000  32 30 32 35 2d 30 31 2d  32 31 54 30 36 3a 35 39  |2025-01-21T06:59|
00000010  3a 33 34 2e 31 33 34 35  32 30 30 31 32 5a        |:34.134520012Z|
0000001e

csi.groupname
value (50 bytes) :
00000000  63 73 69 2d 76 6f 6c 2d  67 72 6f 75 70 2d 30 39  |csi-vol-group-09|
00000010  39 62 61 64 31 38 2d 31  33 32 33 2d 34 64 63 65  |9bad18-1323-4dce|
00000020  2d 62 36 61 30 2d 37 39  30 32 65 31 61 31 63 36  |-b6a0-7902e1a1c6|
00000030  66 30                                             |f0|
00000032

csi.volname
value (47 bytes) :
00000000  76 67 72 63 6f 6e 74 65  6e 74 2d 34 39 66 64 34  |vgrcontent-49fd4|
00000010  35 30 32 2d 63 36 30 30  2d 34 39 32 32 2d 39 33  |502-c600-4922-93|
00000020  31 37 2d 38 32 39 30 36  66 35 31 31 33 32 35     |17-82906f511325|
0000002f
```

Madhu-1 reviewed (commented):
Did we test this change where the VGRClass or VGRContent CRD is not installed in the cluster?

iPraveenParihar commented:
> Did we test this change where the VGRClass or VGRContent CRD is not installed in the cluster? 

If not installed, the container goes into CLBO state. 
```
E0124 10:30:04.728441       1 controller.go:100] failed to start manager failed to wait for vgrcontent-controller caches to sync: timed out waiting for cache to be synced for Kind *v1alpha1.VolumeGroupReplicationContent
E0124 10:30:04.728572       1 cephcsi.go:327] failed to wait for vgrcontent-controller caches to sync: timed out waiting for cache to be synced for Kind *v1alpha1.VolumeGroupReplicationContent
```

I have added a check for required CRD exists 

- If VolumeGroupReplicationContent CRD is not installed
```
I0124 11:40:03.717306       1 cephcsi.go:202] Driver version: canary and Git version: 41a1bc55346ad5fdd03c3c983a8172b9623f6481
I0124 11:40:03.718207       1 cephcsi.go:234] Starting driver type: controller with name: rook-ceph.rbd.csi.ceph.com
E0124 11:40:03.744931       1 volumegroupreplicationcontent.go:107] skipping the VolumeGroupReplicationContent controller: VolumeGroupReplicationContent CRD is not installed: no matches for kind "VolumeGroupReplicationContent" in version "replication.storage.openshift.io/v1alpha1"
```

- If VolumeGroupReplicationClass CRD is not installed
```
E0124 11:37:25.562679       1 volumegroupreplicationcontent.go:107] skipping the VolumeGroupReplicationContent controller: VolumeGroupReplicationClass CRD is not installed: no matches for kind "VolumeGroupReplicationClass" in version "replication.storage.openshift.io/v1alpha1"
```

- If both installed
```
I0124 11:41:09.236050       1 cephcsi.go:202] Driver version: canary and Git version: 41a1bc55346ad5fdd03c3c983a8172b9623f6481
I0124 11:41:09.237333       1 cephcsi.go:234] Starting driver type: controller with name: rook-ceph.rbd.csi.ceph.com
I0124 11:41:10.282007       1 with_retry.go:234] Got a Retry-After 1s response for attempt 1 to https://10.96.0.1:443/apis/replication.storage.openshift.io/v1alpha1/volumegroupreplicationclasses
I0124 11:41:10.288903       1 leaderelection.go:254] attempting to acquire leader lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph...
I0124 11:41:10.303360       1 leaderelection.go:268] successfully acquired lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0124 11:41:10.308395       1 reflector.go:305] Starting reflector *v1.PersistentVolume (10h28m25.882921856s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0124 11:41:10.308450       1 reflector.go:341] Listing and watching *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0124 11:41:10.308600       1 reflector.go:305] Starting reflector *v1alpha1.VolumeGroupReplicationContent (9h48m29.745592474s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0124 11:41:10.308847       1 reflector.go:341] Listing and watching *v1alpha1.VolumeGroupReplicationContent from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0124 11:41:10.312091       1 reflector.go:368] Caches populated for *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0124 11:41:10.313208       1 leaderelection.go:293] successfully renewed lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0124 11:41:10.314264       1 reflector.go:368] Caches populated for *v1alpha1.VolumeGroupReplicationContent from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
```

@Madhu-1, PTAL at recent changes. Thanks!

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,234 @@
+
+func ensureCRDsInstalled(mgr manager.Manager) error {
+	ctx := context.Background()
+	vgrcontentlist := &replicationv1alpha1.VolumeGroupReplicationContentList{}
+	if err := mgr.GetAPIReader().List(ctx, vgrcontentlist); err != nil {
+		if strings.Contains(err.Error(), "no matches for kind") {
+			return fmt.Errorf("%s CRD is not installed: %w", volumeGroupReplicationContentResourceName, err)
+		}
+
+		return fmt.Errorf("failed to list %s resources: %w", volumeGroupReplicationContentResourceName, err)
+	}
+
+	vgrclasslist := &replicationv1alpha1.VolumeGroupReplicationClassList{}
+	if err := mgr.GetAPIReader().List(ctx, vgrclasslist); err != nil {
+		if strings.Contains(err.Error(), "no matches for kind") {
+			return fmt.Errorf("%s CRD is not installed: %w", volumeGroupReplicationClassResourceName, err)
+		}
+
+		return fmt.Errorf("failed to list %s resources: %w", volumeGroupReplicationClassResourceName, err)
+	}
```
we should not be using list can you please check if we can do something like below

```go
_, err := mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
	if err != nil {
		r.Log.Error(err, "error to get resource", "gvk", gvk)
		return
	}
```

Or check ocs-operator or ocs-client-operator code, we check for CRD existence in both operator as well

Madhu-1 commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,234 @@
+	if err = ensureCRDsInstalled(mgr); err != nil {
+		log.ErrorLogMsg("skipping the VolumeGroupReplicationContent controller: %s", err.Error())
+
+		return nil
```
This is returning for all the error messages not for specific error

iPraveenParihar replied:
handled now.

iPraveenParihar commented:
@Madhu-1, I tried your suggestion and here are the testing logs

- VolumeGroupReplicationContent & VolumeGroupReplicationClass CRDs not installed 
```
pm@dhcp53-176:~$ k logs csi-rbdplugin-provisioner-57449f95b4-wr2xq csi-omap-generator
I0127 12:13:12.363090       1 cephcsi.go:202] Driver version: canary and Git version: 41a1bc55346ad5fdd03c3c983a8172b9623f6481
I0127 12:13:12.364133       1 cephcsi.go:234] Starting driver type: controller with name: rook-ceph.rbd.csi.ceph.com
E0127 12:13:12.392704       1 volumegroupreplicationcontent.go:113] Required CRDs (VolumeGroupReplicationContent, VolumeGroupReplicationClass) are not installed
E0127 12:13:12.392760       1 volumegroupreplicationcontent.go:126] Skipping controller creation for VolumeGroupReplicationContent. Please install the missing CRDs.
I0127 12:13:12.393248       1 leaderelection.go:254] attempting to acquire leader lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph...
I0127 12:13:12.409931       1 leaderelection.go:268] successfully acquired lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0127 12:13:12.423799       1 leaderelection.go:293] successfully renewed lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0127 12:13:12.425221       1 reflector.go:305] Starting reflector *v1.PersistentVolume (9h8m38.997192628s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:13:12.425401       1 reflector.go:341] Listing and watching *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:13:12.429495       1 reflector.go:368] Caches populated for *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
```

- Both CRDs installed
```
pm@dhcp53-176:~$ k logs csi-rbdplugin-provisioner-57449f95b4-d284d csi-omap-generator
I0127 12:15:34.451559       1 cephcsi.go:202] Driver version: canary and Git version: 41a1bc55346ad5fdd03c3c983a8172b9623f6481
I0127 12:15:34.452461       1 cephcsi.go:234] Starting driver type: controller with name: rook-ceph.rbd.csi.ceph.com
I0127 12:15:34.481069       1 leaderelection.go:254] attempting to acquire leader lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph...
I0127 12:15:34.499826       1 leaderelection.go:268] successfully acquired lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0127 12:15:34.507273       1 reflector.go:305] Starting reflector *v1.PersistentVolume (10h44m16.47892762s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:15:34.507494       1 reflector.go:341] Listing and watching *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:15:34.510061       1 reflector.go:368] Caches populated for *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:15:34.510225       1 reflector.go:305] Starting reflector *v1alpha1.VolumeGroupReplicationContent (9h46m42.07012007s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:15:34.510267       1 reflector.go:341] Listing and watching *v1alpha1.VolumeGroupReplicationContent from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:15:34.512743       1 reflector.go:368] Caches populated for *v1alpha1.VolumeGroupReplicationContent from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
```

- either one of not installed (VolumeGroupReplicationClass not installed)
```
0127 12:17:24.348264       1 cephcsi.go:202] Driver version: canary and Git version: 41a1bc55346ad5fdd03c3c983a8172b9623f6481
I0127 12:17:24.349930       1 cephcsi.go:234] Starting driver type: controller with name: rook-ceph.rbd.csi.ceph.com
E0127 12:17:24.392450       1 volumegroupreplicationcontent.go:113] Required CRDs (VolumeGroupReplicationClass) are not installed
E0127 12:17:24.392561       1 volumegroupreplicationcontent.go:126] Skipping controller creation for VolumeGroupReplicationContent. Please install the missing CRDs.
I0127 12:17:24.398526       1 leaderelection.go:254] attempting to acquire leader lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph...
I0127 12:17:24.417123       1 leaderelection.go:268] successfully acquired lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0127 12:17:24.430276       1 reflector.go:305] Starting reflector *v1.PersistentVolume (10h22m16.260428995s) from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:17:24.430404       1 reflector.go:341] Listing and watching *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
I0127 12:17:24.431603       1 leaderelection.go:293] successfully renewed lease rook-ceph/rook-ceph.rbd.csi.ceph.com-rook-ceph
I0127 12:17:24.439476       1 reflector.go:368] Caches populated for *v1.PersistentVolume from sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go:108
```

iPraveenParihar committed:
rbd: rename volumeNamePrefix to volumeGroupNamePrefix

Signed-off-by: Praveen M <m.praveen@ibm.com>

Nikhil-Ladha commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
```
nit: 2025

iPraveenParihar replied:
😄 addressed

Nikhil-Ladha commented on internal/controller/volumegroup/volumegroupreplicationcontent.go:
```diff
@@ -0,0 +1,251 @@
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
```
Probably we can optimise it a bit by reducing the number of return statements and if condition like this?
```
        // Trigger reconcile only if the object is not under deletion
        if vgrc.GetDeletionTimestamp().IsZero() {
		err = r.reconcileVGRContent(ctx, vgrc)
	}

	return reconcile.Result{}, err
```

Nikhil-Ladha commented on internal/rbd/manager.go:
```diff
@@ -503,3 +503,134 @@ func (mgr *rbdManager) CreateVolumeGroupSnapshot(
+	}
+	defer vgJournal.Destroy()
+
+	namePrefix = mgr.parameters["groupNamePrefix"]
```
This should be updated to `volumeGroupNamePrefix` too, isn't it?

Nikhil-Ladha reviewed (commented):
Few nits, rest looks good to me

iPraveenParihar commented:
@Nikhil-Ladha, addressed your suggestions. PTAL 😃

Nikhil-Ladha reviewed (approved):
lgtm

Rakshith-R commented on internal/controller/controller.go:
```diff
@@ -20,13 +20,17 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
 	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+
```
```suggestion
```
extra newline

Rakshith-R commented on internal/rbd/manager.go:
```diff
@@ -503,3 +503,134 @@ func (mgr *rbdManager) CreateVolumeGroupSnapshot(
+		}
+		defer func() {
+			if err != nil {
+				err = vgJournal.UndoReservation(ctx, journalPool, vgName, requestName)
+				if err != nil {
+					log.ErrorLog(ctx, "failed to undo the reservation for volume group %q: %w", requestName, err)
+				}
```
Use another internal error variable.

Rakshith-R commented:
@iPraveenParihar there's few ci failures.

Madhu-1 commented:
@mergifyio queue

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.30

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.30

retry, failed to deploy rook and components.

nixpanic commented:
@mergifyio requeue

iPraveenParihar commented:
@Madhu-1, backport to v3.13 is not needed right? Since the VolumeGroupReplication related changes are in v3.13. But not usable since the Ceph support isn't yet available.

Madhu-1 commented:
@iPraveenParihar not required as This is a new feature.
